jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp

```diff
@@ -743,7 +744,634 @@ ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this arg
+{
+  ecma_string_t *appended_string;
+
+  JERRY_ASSERT (start <= end);
+  JERRY_ASSERT (end <= ecma_string_get_length (appended_string_p));
+
+  if (start < end)
+  {
+    ecma_string_t *substring_p = ecma_string_substr (appended_string_p, start, end);
+    appended_string = ecma_concat_ecma_strings (base_string_p, substring_p);
+
+    ecma_deref_ecma_string (substring_p);
+    if (free_base_string)
+    {
+      ecma_deref_ecma_string (base_string_p);
+    }
+  }
+  else if (free_base_string)
+  {
+    appended_string = base_string_p;
+  }
+  else
+  {
+    appended_string = ecma_copy_or_ref_ecma_string (base_string_p);
+  }
+
+  return appended_string;
+} /* ecma_builtin_string_prototype_object_replace_append_substr */
+
+/**
+ * Generic helper function to perform the find the next match
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_string_prototype_object_replace_match (replace_search_context *context_p) /**< search context */
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  context_p->match_start = 0;
+  context_p->match_end = 0;
+
+  if (context_p->is_regexp)
+  {
+    ecma_value_t exec_arguments[1] = { context_p->input_string };
+
+    ECMA_TRY_CATCH (match_value,
+                    ecma_builtin_regexp_prototype_dispatch_routine (LIT_MAGIC_STRING_EXEC,
+                                                                    context_p->regexp_or_search_string,
+                                                                    exec_arguments,
+                                                                    1),
+                    ret_value);
+
+    if (!ecma_is_value_null (match_value))
+    {
+      JERRY_ASSERT (ecma_is_value_object (match_value));
+
+      ecma_object_t *match_object_p = ecma_get_object_from_value (match_value);
+      ecma_string_t *index_string_p = ecma_get_magic_string (LIT_MAGIC_STRING_INDEX);
+      ecma_string_t *zero_string_p = ecma_new_ecma_string_from_uint32 (0);
+
+      ECMA_TRY_CATCH (index_value,
+                      ecma_op_object_get (match_object_p, index_string_p),
+                      ret_value);
+
+      ECMA_TRY_CATCH (result_string_value,
+                      ecma_op_object_get (match_object_p, zero_string_p),
+                      ret_value);
+
+      /* We directly call the built-in exec, so
+       * we can trust in the returned value. */
+
+      JERRY_ASSERT (ecma_is_value_number (index_value));
+      JERRY_ASSERT (ecma_is_value_string (result_string_value));
+
+      /* We use the length of the result string to determine the
+       * match end. This works regardless the global flag is set. */
+      ecma_string_t *result_string_p = ecma_get_string_from_value (result_string_value);
+
+      context_p->match_start = (ecma_length_t) (*ecma_get_number_from_value (index_value));
```

Comment:
@zherczeg, I think adding an assert would be ok for the case. If `index_num` can be represented in `ecma_length_t`, then simple C-style conversion would work correctly. Maybe, later we should add interface for the conversion that would perform the check internally.

P.S. Sorry for deleting my first comment to the code block (about "would be the conversion correct if `ecma_length_t` would be, for example, `uint16_t`"). I've deleted the comment before I've seen your answer https://github.com/Samsung/jerryscript/pull/474#discussion_r35731210, and added new comment https://github.com/Samsung/jerryscript/pull/474#discussion_r35731390.